### ISSUE TITLE
How to make the pixel buffer larger than the window

### ISSUE THREAD

**Issue by wiz21b** [parasyte/pixels#319]
Hello,

I've been using pixels for a while now and I'm now stuck at some edge case. Basically, I'd like to put a bit more information in the pixels buffer than what will be actually drawn. This ends up with my "buffer" being 1 pixel wider than the actual window's width. In the documentation of the SurfaceTexture, it is written : "It is recommended (but not required) that the width and height are equivalent to the physical dimensions of the surface.". But that's not 100% clear to me what happens behind the scenes. Of course I have tried to make a surface texture wider than the actual window size but this brings a white picture (I was expecting a distorted picture, maybe zoomed). I thought that maybe the scaling matrix was at fault, esp. the constraint over the scale factor, so I set it to one. But that didn't succeed...

If you could just tell me where I should start looking that would be very kind. I've also looked in other places in the code and saw there was a "render texture" and a "backing texture" but I fail to understand the difference.

Thank your for your time...

**Comment by parasyte** [comment]
This should give you a much clearer sense of what the two textures look like. I captured these with [RenderDoc](https://renderdoc.org/) on Windows 11.

This is the "backing texture" for the `minimal-winit` example. It's exactly what you expect your pixel data to look like.

![backing-texture](https://user-images.githubusercontent.com/456942/206027847-c99a3071-07f4-476a-94ae-30186a96c9c5.png)

And this one is the "render texture". Both textures were captured on the same frame, and the window was resized slightly bigger than the default (to give it a border). This texture is "what you actually see" in the window.

![render-texture](https://user-images.githubusercontent.com/456942/206028020-f5ea5459-0c8c-4b7f-bdb1-93ce12d5a2c8.png)

**Comment by wiz21b** [comment]
>  The only thing I can think of is that you might be trying to code-golf to reduce a few lines of code for copying a small region from a larger texture. If that is the case, we have better ways...

Yeah, I told you it was "edge case" so yes, code golf. I wanted to reduce the number of texture uploads :-) To explain it fully: my main texture is a composite PAL/NTSC signal. As you know there's a notion of "color burst" which tells how to decode the composite signal: that is with or without colour. I wanted to put the colour bust on/off in my texture, on its right edge "a bit like" what is done with a TV. If I don't do that, then I have to put that information in another texture which means more code and more GPU transfers... I wanted to avoid that :-) So yes, definitely code golf (but not sub-par, I admit)

**Comment by wiz21b** [comment]
For the backing/render texture... If it's to get a border, why didn't you use a regular texture border, this would have avoided the additional texture. But sure, you knwo what you do better than I :-)

**Comment by parasyte** [comment]
I use the [`pix`](https://docs.rs/pix/latest/pix/) crate with `pixels` in a game I am working on to handle compositing operations. It was a little work to setup initially, but it allows copying small regions from any offset of the source to any offset in the destination as shown in [the documentation](https://docs.rs/pix/latest/pix/struct.Raster.html#method.copy_raster):

```
*------------+      *-------------+
|            |      |    *------+ |
| *------+   |      |    |      | |
| |      |   |      |    | from | |
| |  to  |   | <--- |    +------+ |
| +------+   |      |             |
|            |      |     src     |
|    self    |      +-------------+
+------------+
```

This kind of compositing is almost certainly what you want to do, instead of trying to clip the surface texture. I don't know about other crates, but with `pix` you can do the copy from a larger image (your PAL/NTSC simulated video) to a smaller image (the pixel buffer managed by `pixels`) without any unnecessary copies or texture uploads.

> For the backing/render texture... If it's to get a border, why didn't you use a regular texture border, this would have avoided the additional texture. But sure, you knwo what you do better than I :-)

What is a "regular texture border"? Regardless, you always need at least two textures. The source texture that you upload to, and the destination render target that the shader writes to. The render target has to be window-sized (for reasons described above). Scaling and adding a border are just two things that are done for convenient/nice handling of the size difference between the two textures.

**Comment by wiz21b** [comment]
hmmm... I forgot to mention that my conversion from "pixels" buffer (composite signal) to the displayed RGB signal is done with  a WGSL shader. Hence my willingness to keep the pipeline simple by not multiplying GPU textures. In my case, the "color burst" signal is an information I use to change the behavior of the shader (which also goes in the direction of having everything crammed into a single texture -- although not 100% needed).

For the regular texture border, I was refering to the fact that one can tell a shader to use a given border color when it accesses texels which coordinates are outside the texture (that's part of the choice to wrap the texture, clamp it, mirror it, etc.)

**Comment by parasyte** [comment]
That is an important detail, yeah. I was under the impression that the "analog video signal" was constructed on the CPU side. But then the good news is that you should be able to do the clipping in your shader, right?

> For the regular texture border, I was refering to the fact that one can tell a shader to use a given border color when it accesses texels which coordinates are outside the texture (that's part of the choice to wrap the texture, clamp it, mirror it, etc.)

I follow. You can see that the pipeline clamps the texture coordinates on the edges: https://github.com/parasyte/pixels/blob/bc8235fdb10fc7ea20190d27a8040e36fa437f8f/src/renderers.rs#L34-L36 This is the reason the border is composed of the clear color, instead of, say, wrapping the texture.
